### PR TITLE
[CT-1361] Add Support for TWAP Events in the Indexer

### PR DESCRIPTION
**Pull request by anmolagrawal345**
### Changelist
This PR is dependent on #2780, #2785, and #2787.

This change focuses on the indexer side of the changes for the TWAP order project. Now that the protocol can handle TWAP orders, we need emit and handle new events to add support to the indexer. The general workflow is going to be:
1. TWAP Order emits an order placement message and added to orders table
2. Suborders fills emit messages but placement does not (TBD)
3. Add a handler on the indexer side to update the quantity and average price on suborder fill messages

### Test Plan
Extend existing order placement handler tests to support twap orders and do e2e testing in dev/staging.

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

## Summary by CodeRabbit

* **New Features**
  * Added TWAP order support (TWAP orders and TWAP suborders) across placement, fills, removals, and public APIs; responses expose optional duration, interval, priceTolerance.
* **Database**
  * Added orders columns for duration, interval, priceTolerance and extended order/fill enums to include TWAP types; migration included.
* **Documentation**
  * OpenAPI and public API docs updated to describe TWAP types, fields, and query filtering.
* **Tests**
  * Extended test suites and helpers to cover TWAP workflows and edge cases.
* **Chores**
  * Added DB helper constants and SQL mapping for TWAP flag handling.

### CODE REVIEW

File: indexer/packages/postgres/src/lib/protocol-translations.ts [hunk 1]
```diff
@@ -318,12 +321,32 @@ export function getGoodTilBlockTime(order: IndexerOrder): IsoString | undefined
...
+    case 128:
+      return OrderType.TWAP;
+    case 256:
+      return OrderType.TWAP_SUBORDER;
+    default:
+      throw new Error(`Unexpected OrderFlags: ${orderFlag}`);
   }
-
-  return CONDITION_TYPE_TO_ORDER_TYPE_MAP[protocolConditionType];
 }
```
[the remark this answers is not in the document]
Comment (jusbar23) [reply indexer/packages/postgres/src/lib/protocol-translations.ts]: Can you apply this change?

File: indexer/packages/postgres/src/models/order-model.ts [hunk 2]
```diff
@@ -96,6 +96,9 @@ export default class OrderModel extends BaseModel {
         triggerPrice: { type: ['string', 'null'], default: null, pattern: NonNegativeNumericPattern },
         updatedAt: { type: 'string', format: 'date-time' },
         updatedAtHeight: { type: 'string', pattern: IntegerPattern },
+        duration: { type: ['string', 'null'], default: null, pattern: NonNegativeNumericPattern },
+        interval: { type: ['string', 'null'], default: null, pattern: NonNegativeNumericPattern },
+        priceTolerance: { type: ['string', 'null'], default: null, pattern: NonNegativeNumericPattern },
```
[the remark this answers is not in the document]
Comment (jusbar23) [reply indexer/packages/postgres/src/models/order-model.ts:108]: Please fix

File: indexer/services/comlink/src/controllers/api/v4/fills-controller.ts [hunk 3]
```diff
@@ -137,27 +139,34 @@ class FillsController extends Controller {
       },
     );
 
+    // = await FillTable.findAll(
+    //   {
```
Comment (jusbar23) [inline indexer/services/comlink/src/controllers/api/v4/fills-controller.ts]: Delete this

File: indexer/services/ender/__tests__/helpers/indexer-proto-helpers.ts [hunk 4]
```diff
@@ -451,6 +457,17 @@ export function createOrder({
     };
   }
 
+  if (duration !== undefined && interval !== undefined && priceTolerance !== undefined) {
```
Comment (jusbar23) [inline indexer/services/ender/__tests__/helpers/indexer-proto-helpers.ts:462]: So when TWAP is enabled, all of these fields need to be set right? Should we throw an error if this condition is only partially met?

File: indexer/services/ender/src/scripts/handlers/dydx_stateful_order_handler.sql [hunk 5]
```diff
@@ -117,6 +138,8 @@ BEGIN
                 order_record."status" = 'CANCELED';
         END CASE;
 
+        RAISE WARNING 'RECEIVED ORDER REMOVAL/CANCEL EVENT: % | UUID: %', event_data, dydx_uuid_from_order_id(order_id);
```
Comment (jusbar23) [inline indexer/services/ender/src/scripts/handlers/dydx_stateful_order_handler.sql]: As we discussed offline: Can we add a way to deactivate warnings in production?

File: indexer/services/ender/src/scripts/helpers/dydx_get_order_status.sql [hunk 6]
```diff
@@ -30,7 +30,7 @@ BEGIN
     IF total_filled >= size THEN
         RETURN 'FILLED';
     /** Order flag of 64 is a stateful term order */
-    ELSIF order_flags = 64 THEN /** 1. Stateful Order */
+    ELSIF order_flags = 64 OR order_flags = 256 THEN /** 1. Stateful Order or Twap Suborder */
```
Comment (jusbar23) [inline indexer/services/ender/src/scripts/helpers/dydx_get_order_status.sql]: These magic numbers are going to get confusing, can we make these constants?

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql [hunk 7]
```diff
@@ -89,7 +89,7 @@
     builder_address = dydx_get_builder_address(fill_liquidity, event_data);
     affiliate_rev_share = dydx_trim_scale(dydx_from_jsonlib_long(event_data->'affiliateRevShare') *
                                     power(10, asset_record."atomicResolution")::numeric);
-
+    fill_type = CASE WHEN (order_->'orderId'->>'orderFlags')::bigint = 256 THEN 'TWAP_SUBORDER' ELSE fill_type END;
```
[the remark this answers is not in the document]
Comment (jusbar23) [reply indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql]: Is this case tested?

File: indexer/services/ender/src/validators/stateful-order-validator.ts [hunk 8]
```diff
@@ -32,14 +33,20 @@ export class StatefulOrderValidator extends Validator<StatefulOrderEventV1> {
       this.event.orderRemoval === undefined &&
       this.event.conditionalOrderPlacement === undefined &&
       this.event.conditionalOrderTriggered === undefined &&
-      this.event.longTermOrderPlacement === undefined
+      this.event.longTermOrderPlacement === undefined &&
+      this.event.twapOrderPlacement === undefined
```
Comment (jusbar23) [inline indexer/services/ender/src/validators/stateful-order-validator.ts]: We should move this into a separate function tbh, this conditional is getting long

File: indexer/services/ender/src/validators/stateful-order-validator.ts [hunk 9]
```diff
@@ -32,14 +33,20 @@
...
-        'longTermOrderPlacement must be defined in StatefulOrderEvent',
+        'longTermOrderPlacement, or twapOrderPlacement must be defined in StatefulOrderEvent',
         { event: this.event },
       );
     }
+    logger.info({
+      message: 'StatefulOrderValidator',
+      at: 'validate',
+      event: this.event,
+    });
```
[the remark this answers is not in the document]
Comment (jusbar23) [reply indexer/services/ender/src/validators/stateful-order-validator.ts]: Remove this

File: indexer/services/ender/src/validators/stateful-order-validator.ts [hunk 10]
```diff
@@ -48,9 +55,9 @@ export class StatefulOrderValidator extends Validator<StatefulOrderEventV1> {
       this.validateConditionalOrderPlacement(this.event.conditionalOrderPlacement);
     } else if (this.event.conditionalOrderTriggered !== undefined) {
       this.validateConditionalOrderTriggered(this.event.conditionalOrderTriggered);
-    } else { // longTermOrderPlacement
+    } else if (this.event.longTermOrderPlacement !== undefined) {
       this.validateLongTermOrderPlacement(this.event.longTermOrderPlacement!);
-    }
+    } // validate twap?
```
Comment (jusbar23) [inline indexer/services/ender/src/validators/stateful-order-validator.ts]: nit pick: question mark

File: protocol/indexer/events/stateful_order.go [hunk 11]
```diff
@@ -63,3 +63,17 @@ func NewConditionalOrderTriggeredEvent(
 		},
 	}
 }
+
+func NewTwapOrderPlacementEvent(
+	order clobtypes.Order,
+) *StatefulOrderEventV1 {
+	indexerOrder := v1.OrderToIndexerOrder(order)
```
Comment (jusbar23) [inline protocol/indexer/events/stateful_order.go:70]: Nit pick: Just pass this in directly? No need to create a variable

File: indexer/packages/postgres/src/models/order-model.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jusbar23) [reply indexer/packages/postgres/src/models/order-model.ts:108]: Do you need to add it to the SQLtoJsonConversion too?

File: indexer/services/ender/src/scripts/handlers/dydx_stateful_order_handler.sql [hunk 12]
```diff
@@ -21,10 +21,19 @@ DECLARE
     perpetual_market_record perpetual_markets%ROWTYPE;
     order_record orders%ROWTYPE;
     subaccount_record subaccounts%ROWTYPE;
+    order_flag bigint;
 BEGIN
     /** TODO(IND-334): Remove after deprecating StatefulOrderPlacementEvent. */
-    IF event_data->'orderPlace' IS NOT NULL OR event_data->'longTermOrderPlacement' IS NOT NULL OR event_data->'conditionalOrderPlacement' IS NOT NULL THEN
-        order_ = coalesce(event_data->'orderPlace'->'order', event_data->'longTermOrderPlacement'->'order', event_data->'conditionalOrderPlacement'->'order');
+    IF event_data->'orderPlace' IS NOT NULL OR event_data->'longTermOrderPlacement' IS NOT NULL OR event_data->'conditionalOrderPlacement' IS NOT NULL OR event_data->'twapOrderPlacement' IS NOT NULL THEN
```
Comment (jusbar23) [inline indexer/services/ender/src/scripts/handlers/dydx_stateful_order_handler.sql]: Why don't you coalesce these values and then just check if the resulting value is NULL to go down the branch?

File: indexer/services/ender/src/scripts/helpers/dydx_uuid_from_order_id.sql [hunk 13]
```diff
@@ -5,6 +5,20 @@ CREATE OR REPLACE FUNCTION dydx_uuid_from_order_id(order_id jsonb) RETURNS uuid
...
+    RAISE WARNING 'MATCH TWAP SUBORDER: %', (order_id->>'orderFlags')::bigint = 256;
+
+    IF (order_id->>'orderFlags')::bigint = 256 THEN
+      RAISE WARNING 'MAPPING TWAP SUBORDER TO PARENT ORDER IN UUID CHECK';
+        -- Twap suborders should be mapped to their parent order.
+        return dydx_uuid_from_order_id_parts(
+            dydx_uuid_from_subaccount_id(order_id->'subaccountId'),
+            order_id->>'clientId',
+            order_id->>'clobPairId',
+            '128');
```
Comment (jusbar23) [inline indexer/services/ender/src/scripts/helpers/dydx_uuid_from_order_id.sql]: Nit: These should be coded as constants

File: indexer/services/ender/__tests__/helpers/indexer-proto-helpers.ts (hunk 4 above)
Comment (anmolagrawal345) [reply indexer/services/ender/__tests__/helpers/indexer-proto-helpers.ts:462]: We check these values in stateless validation in message_place_order so think we should be fine here

File: indexer/services/ender/src/scripts/handlers/dydx_stateful_order_handler.sql (hunk 5 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/handlers/dydx_stateful_order_handler.sql]: these logs are a bit excessive regardless so going to remove them before merging. we can look into adding debug logging in a subsequent change?

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql]: added test coverage to the fill type in order-handler.test

File: indexer/services/ender/src/scripts/helpers/dydx_uuid_from_order_id.sql (hunk 13 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_uuid_from_order_id.sql]: created a constants schema to store the order flags. just modified twap use cases for now to keep this change granular

File: indexer/services/ender/src/validators/stateful-order-validator.ts (hunk 10 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/validators/stateful-order-validator.ts]: added validateTwapOrderPlacement

File: protocol/indexer/events/stateful_order.go (hunk 11 above)
Comment (anmolagrawal345) [reply protocol/indexer/events/stateful_order.go:70]: tbh would leave for readability here where we're moving from protocol to indexer models

File: indexer/packages/postgres/src/lib/protocol-translations.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (anmolagrawal345) [reply indexer/packages/postgres/src/lib/protocol-translations.ts]: done

File: indexer/packages/postgres/src/models/order-model.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (anmolagrawal345) [reply indexer/packages/postgres/src/models/order-model.ts:108]: done

File: indexer/packages/postgres/src/db/migrations/migration_files/20250428122430_add_twap_types_to_order_table.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+            'TAKE_PROFIT',
+            'TAKE_PROFIT_MARKET',
+            'LIQUIDATED',
+            'LIQUIDATION',
+            'HARD_TRADE',
+            'FAILED_HARD_TRADE',
+            'TRANSFER_PLACEHOLDER',
+          ],
+    ));
+}
```
[the remark this answers is not in the document]
Comment (anmolagrawal345) [reply indexer/packages/postgres/src/db/migrations/migration_files/20250428122430_add_twap_types_to_order_table.ts]: // copied from this answer: https://github.com/knex/knex/issues/1699
// Generates raw SQL to updating check constraints on an enum column with minimal locking.
export function formatAlterTableEnumSql(
  tableName: string,
  columnName: string,
  enums: string[],
): string {
  const constraintName = `${tableName}_${columnName}_check`;
  return [
    `ALTER TABLE "${tableName}" DROP CONSTRAINT IF EXISTS "${constraintName}";`,
    `ALTER TABLE "${tableName}" ADD CONSTRAINT "${constraintName}" CHECK ("${columnName}" = ANY (ARRAY['${enums.join(
      '\'::text, \'',
    )}'::text])) NOT VALID;`,
  ].join('\n');
}

this is the implementation of the formatAlterTableEnumSql function which should drop an existing constraint and create a new one on the new values for the column

File: indexer/packages/postgres/src/db/migrations/migration_files/20250428122430_add_twap_types_to_order_table.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import * as Knex from 'knex';
+
+import { formatAlterTableEnumSql } from '../helpers';
+
+export async function up(knex: Knex): Promise<void> {
+  return knex.raw(formatAlterTableEnumSql(
+    'orders',
+    'type',
+    [
```
Comment (shrenujb) [inline indexer/packages/postgres/src/db/migrations/migration_files/20250428122430_add_twap_types_to_order_table.ts]: Are these new types or already existing types? What field was added for this to be changed?

File: indexer/packages/postgres/src/db/migrations/migration_files/20250715122430_add_twap_suborder_type_to_fills_table.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
+import * as Knex from 'knex';
+
+import { formatAlterTableEnumSql } from '../helpers';
+
+export async function up(knex: Knex): Promise<void> {
+  return knex.raw(formatAlterTableEnumSql(
+    'fills',
+    'type',
+    ['LIMIT', 'LIQUIDATED', 'LIQUIDATION', 'DELEVERAGED', 'OFFSETTING', 'TWAP_SUBORDER'],
```
Comment (shrenujb) [inline indexer/packages/postgres/src/db/migrations/migration_files/20250715122430_add_twap_suborder_type_to_fills_table.ts]: Same question as above

File: protocol/x/clob/keeper/twap_order_state.go [hunk 17]
```diff
@@ -278,6 +278,7 @@ func (k Keeper) UpdateTWAPOrderRemainingQuantityOnFill(
 	twapOrderPlacementBytes := k.cdc.MustMarshal(&twapOrderPlacement)
 	store.Set(orderKey, twapOrderPlacementBytes)
 
+	// would be cleaner to send indexer event here
```
Comment (shrenujb) [inline protocol/x/clob/keeper/twap_order_state.go]: Is this a TODO? Do you want to change this now or do this later?

File: proto/dydxprotocol/indexer/protocol/v1/clob.proto [hunk 18]
```diff
@@ -198,3 +201,20 @@ enum ClobPairStatus {
   // short-term orders are left to expire.
   CLOB_PAIR_STATUS_FINAL_SETTLEMENT = 6;
 }
+
+// TwapParameters represents the necessary configuration for a TWAP order.
+message TwapParameters {
```
Comment (shrenujb) [inline proto/dydxprotocol/indexer/protocol/v1/clob.proto]: Why do we need to add this separately to the clob proto?

File: indexer/packages/postgres/src/db/migrations/migration_files/20250428122430_add_twap_types_to_order_table.ts (hunk 15 above)
Comment (anmolagrawal345) [reply indexer/packages/postgres/src/db/migrations/migration_files/20250428122430_add_twap_types_to_order_table.ts]: TWAP and TWAP_SUBORDER were added here. The formatAlterTableEnumSql function essentially just creates a new constraint on the column, so you need to pass in every value of the type.

File: indexer/packages/postgres/src/db/migrations/migration_files/20250715122430_add_twap_suborder_type_to_fills_table.ts (hunk 16 above)
Comment (anmolagrawal345) [reply indexer/packages/postgres/src/db/migrations/migration_files/20250715122430_add_twap_suborder_type_to_fills_table.ts]: TWAP_SUBORDER was the fill type that was added

File: protocol/x/clob/keeper/twap_order_state.go (hunk 17 above)
Comment (anmolagrawal345) [reply protocol/x/clob/keeper/twap_order_state.go]: deleted comment - it was just for updating the parent twap state in the sql handlers but that's been handled now

File: proto/dydxprotocol/indexer/protocol/v1/clob.proto (hunk 18 above)
Comment (anmolagrawal345) [reply proto/dydxprotocol/indexer/protocol/v1/clob.proto]: ah I actually don't I imported the messages from order.proto now - and also updated BuilderCodeParameters to import from the same

File: protocol/x/clob/keeper/twap_order_state.go [hunk 19]
```diff
@@ -398,3 +398,30 @@ func (k Keeper) GenerateSuborder(
 
 	return &order, true
 }
+
+// safeHandleMsgPlaceOrder safely calls HandleMsgPlaceOrder with panic recovery.
+// This is used in end blockers where panics should be caught and logged rather than
+// causing the entire block to fail.
+func (k Keeper) safeHandleMsgPlaceOrder(
```
Comment (shrenujb) [inline protocol/x/clob/keeper/twap_order_state.go]: Feel like this should be part of a different PR given its important and is not indexer related

File: protocol/x/clob/keeper/twap_order_state.go [hunk 20]
```diff
@@ -398,3 +398,30 @@ func (k Keeper) GenerateSuborder(
...
+	ctx sdk.Context,
+	msg *types.MsgPlaceOrder,
+	isStateful bool,
+) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
```
Comment (shrenujb) [inline protocol/x/clob/keeper/twap_order_state.go]: What does this do exactly?

File: indexer/services/ender/src/scripts/helpers/dydx_get_order_status.sql (hunk 6 above)
Comment (shrenujb) [reply indexer/services/ender/src/scripts/helpers/dydx_get_order_status.sql]: Agreed with Justin here

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql [hunk 21]
```diff
@@ -95,6 +95,8 @@ BEGIN
                                     power(10, asset_record."atomicResolution")::numeric);
     order_router_address = dydx_get_order_router_address(fill_liquidity, event_data);
 
+    fill_type = CASE WHEN (order_->'orderId'->>'orderFlags')::bigint = constants.order_flag_twap_suborder() THEN 'TWAP_SUBORDER' ELSE fill_type END;
```
Comment (shrenujb) [inline indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql:98]: Why is the filltype 'TWAP_SUBORDER' and not just 'TWAP'?

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql [hunk 22]
```diff
@@ -116,44 +117,85 @@ BEGIN
...
+        IF jsonb_extract_path(order_, 'orderId', 'orderFlags')::bigint = constants.order_flag_twap_suborder() THEN
+            order_record."price" = dydx_get_weighted_average(order_record."price", order_record."totalFilled", maker_price, fill_amount);
+            order_record."totalFilled" = order_record."totalFilled" + fill_amount; 
+
+            order_record."status" = dydx_get_order_status(order_record."totalFilled", order_record."size", order_canceled_status, jsonb_extract_path(order_, 'orderId', 'orderFlags')::bigint, order_record."timeInForce");
+            
+            -- Twap suborders shouldn't update all the fields. For example, order flags should remain 128 (not updated to 256).
+            UPDATE orders
+            SET
+                "side" = order_record."side",
```
Comment (shrenujb) [inline indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql]: why would side be updated?

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql [hunk 23]
```diff
@@ -116,44 +117,85 @@ BEGIN
...
+                "status" = order_record."status",
+                "price" = order_record."price",
+                "reduceOnly" = order_record."reduceOnly",
```
Comment (shrenujb) [inline indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql]: why would reduceOnly be updated as a result of a fill?

File: indexer/services/ender/src/scripts/helpers/dydx_protocol_condition_type_to_order_type.sql [hunk 24]
```diff
@@ -1,17 +1,25 @@
...
   (Note that no text should exist before the function declaration to ensure that exception line numbers are correct.)
 */
 BEGIN
-    CASE condition_type
-        WHEN '-1'::jsonb THEN RETURN 'LIMIT'; /** UNRECOGNIZED */
-        WHEN '0'::jsonb THEN RETURN 'LIMIT'; /** CONDITION_TYPE_UNSPECIFIED */
-        WHEN '1'::jsonb THEN RETURN 'STOP_LIMIT'; /** CONDITION_TYPE_STOP_LOSS */
-        WHEN '2'::jsonb THEN RETURN 'TAKE_PROFIT'; /** CONDITION_TYPE_TAKE_PROFIT */
-        ELSE RAISE EXCEPTION 'Unexpected ConditionType: %', condition_type;
+    CASE order_flags
```
Comment (shrenujb) [inline indexer/services/ender/src/scripts/helpers/dydx_protocol_condition_type_to_order_type.sql]: this function seemed specific to condition type. Why did we also add the twap order flags here?

File: protocol/x/clob/keeper/twap_order_state.go (hunk 20 above)
Comment (anmolagrawal345) [reply protocol/x/clob/keeper/twap_order_state.go]: from my understanding calling recover here stops a panicking goroutine and fetches the error - so if something went wrong with handlePlaceOrder we would catch it here

File: protocol/x/clob/keeper/twap_order_state.go (hunk 19 above)
Comment (anmolagrawal345) [reply protocol/x/clob/keeper/twap_order_state.go]: I can move it if we prefer - I think we talked offline about it when we got the audit and I mentioned I would just add it here since it was a small change

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql (hunk 21 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql:98]: Because the fills are each individually twap suborders whereas the parent order is tagged as twap

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql (hunk 22 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql]: removed

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql (hunk 23 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql]: agreed - removed

File: indexer/services/ender/src/scripts/helpers/dydx_get_order_status.sql (hunk 6 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_get_order_status.sql]: updated using the constants

File: protocol/x/clob/keeper/twap_order_state.go (hunk 19 above)
Comment (shrenujb) [reply protocol/x/clob/keeper/twap_order_state.go]: Was this recommended by Informal?

File: indexer/services/ender/src/scripts/helpers/dydx_protocol_condition_type_to_order_type.sql (hunk 24 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_protocol_condition_type_to_order_type.sql]: well not really because it was checking for a -1 condition type to just set the type to limit, so now I just explicitly check the order flag now to decide what the order type is. I think this was historically done because we default our order types to LIMIT, but now that thats not the case I think checking against order flags here made sense

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql (hunk 21 above)
Comment (shrenujb) [reply indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql:98]: Right but you can't really have fills for the main TWAP order and fills for all twap suborders are still TWAP fills. Anyways, this was more of a comment around nomenclature. I dont have strong opinions on this

File: protocol/x/clob/keeper/twap_order_state.go (hunk 20 above)
Comment (shrenujb) [reply protocol/x/clob/keeper/twap_order_state.go]: Can you add a code comment with this explanation?

File: protocol/x/clob/keeper/twap_order_state.go (hunk 19 above)
Comment (shrenujb) [reply protocol/x/clob/keeper/twap_order_state.go]: Yes can we please move it to a different PR. Don't wanna intermix things if we ever need to revert some changes

File: indexer/services/ender/src/scripts/helpers/dydx_protocol_condition_type_to_order_type.sql (hunk 24 above)
Comment (shrenujb) [reply indexer/services/ender/src/scripts/helpers/dydx_protocol_condition_type_to_order_type.sql]: In that case, this function should be renamed since it doesn't purely just use condition_type anymore
Something like "convert_to_order_type"

File: indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql (hunk 21 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_order_fill_handler_per_order.sql:98]: right fair I am just of the opinion that SUBORDER is more clear

File: protocol/x/clob/keeper/twap_order_state.go (hunk 20 above)
Comment (anmolagrawal345) [reply protocol/x/clob/keeper/twap_order_state.go]: done

File: protocol/x/clob/keeper/twap_order_state.go [hunk 25]
```diff
@@ -357,6 +373,17 @@ func (k Keeper) calculateSuborderQuantums(
...
+	reason := indexershared.OrderRemovalReason_ORDER_REMOVAL_REASON_UNSPECIFIED
+	if errors.Is(err, types.ErrStatefulOrderCollateralizationCheckFailed) {
+		reason = indexershared.OrderRemovalReason_ORDER_REMOVAL_REASON_UNDERCOLLATERALIZED
+	} else if errors.Is(err, types.ErrWouldViolateIsolatedSubaccountConstraints) {
+		reason = indexershared.OrderRemovalReason_ORDER_REMOVAL_REASON_VIOLATES_ISOLATED_SUBACCOUNT_CONSTRAINTS
+	}
+
+	return reason
+}
+
```
[the remark this answers is not in the document]
Comment (anmolagrawal345) [reply protocol/x/clob/keeper/twap_order_state.go:385]: I think this is wrong - errorsmod.Is is not defined? Also errors.Is is used in several other places like memclob.go

File: indexer/services/ender/src/scripts/helpers/dydx_protocol_condition_type_to_order_type.sql (hunk 24 above)
Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_protocol_condition_type_to_order_type.sql]: done renamed to dydx_protocol_convert_to_order_type

File: protocol/x/clob/keeper/twap_order_state.go [hunk 26]
```diff
@@ -206,11 +212,23 @@ func (k Keeper) GenerateAndPlaceTriggeredTwapSuborders(ctx sdk.Context) {
 			)
 
 			// place triggered suborder
-			err := k.HandleMsgPlaceOrder(ctx, &types.MsgPlaceOrder{Order: *op.suborderToPlace}, true)
+			err := k.safeHandleMsgPlaceOrder(ctx, &types.MsgPlaceOrder{Order: *op.suborderToPlace}, true)
 			if err != nil {
 				// TODO: (anmol) emit indexer event (TWAP error)
```
Comment (shrenujb) [inline protocol/x/clob/keeper/twap_order_state.go]: Remove this todo?

File: protocol/x/clob/keeper/twap_order_state.go [hunk 27]
```diff
@@ -206,11 +212,23 @@ func (k Keeper) GenerateAndPlaceTriggeredTwapSuborders(ctx sdk.Context) {
 			)
 
 			// place triggered suborder
-			err := k.HandleMsgPlaceOrder(ctx, &types.MsgPlaceOrder{Order: *op.suborderToPlace}, true)
+			err := k.safeHandleMsgPlaceOrder(ctx, &types.MsgPlaceOrder{Order: *op.suborderToPlace}, true)
```
Comment (shrenujb) [inline protocol/x/clob/keeper/twap_order_state.go:214]: this needs to be rebased?

Comment (anmolagrawal345) [reply protocol/x/clob/keeper/twap_order_state.go:214]: rebased

File: indexer/services/ender/src/scripts/helpers/dydx_liquidation_fill_handler_per_order.sql [hunk 28]
```diff
@@ -164,11 +180,31 @@ BEGIN
             order_record."totalFilled" = fill_amount;
             order_record."status" = dydx_get_order_status(fill_amount, order_size, 'NOT_CANCELED', order_record."orderFlags", order_record."timeInForce");
             order_record."createdAtHeight" = block_height;
+
+            IF jsonb_extract_path(order_, 'orderId', 'orderFlags')::bigint = constants.order_flag_twap_suborder() THEN
```
Comment (shrenujb) [inline indexer/services/ender/src/scripts/helpers/dydx_liquidation_fill_handler_per_order.sql:184]: We should never expect a fill from twap suborder to not have an entry for the parent?

Comment (anmolagrawal345) [reply indexer/services/ender/src/scripts/helpers/dydx_liquidation_fill_handler_per_order.sql:184]: discussed offline and added warnings

**Review by shrenujb — APPROVED** [review]

File: indexer/packages/postgres/src/db/migrations/migration_files/20250423144330_add_twap_fields_to_orders_and_fills_table.ts [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+      'TRANSFER_PLACEHOLDER',
+    ],
+  ));
+
+  await knex.raw(formatAlterTableEnumSql(
+    'fills',
+    'type',
+    ['LIMIT', 'LIQUIDATED', 'LIQUIDATION', 'DELEVERAGED', 'OFFSETTING'],
+  ));
+}
```
[the remark this answers is not in the document]
Comment (anmolagrawal345) [reply indexer/packages/postgres/src/db/migrations/migration_files/20250423144330_add_twap_fields_to_orders_and_fills_table.ts:71]: tested this locally and it seems that the down migration works fine as is. the new constraint thats created applies only to new values being inserted, so existing TWAP fields should not be a problem.

**Comment by anmolagrawal345** [comment]
@Mergifyio backport release/protocol/v9.x

**Comment by anmolagrawal345** [comment]
@Mergifyio backport release/indexer/v9.x
